### PR TITLE
add case live_backup_add_bitmap

### PR DESCRIPTION
**Pull request by lolyu**
Aims to test bitmap addition for qcow2/raw images with persistent
on/off.
Signed-off-by: lolyu <lolyu@redhat.com>

### CODE REVIEW

**Comment by lolyu** [comment]
```
(01/12) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (73.18 s)
 (02/12) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (74.07 s)
 (03/12) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (72.59 s)
 (04/12) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_on: PASS (45.59 s)
 (05/12) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (69.11 s)
 (06/12) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (72.87 s)
```

**Comment by lolyu** [comment]
@aliang123 please review.

**Comment by lolyu** [comment]
depends on [](https://github.com/avocado-framework/avocado-vt/pull/2100)

**Comment by lolyu** [comment]
> For # ensure qmp support part, we have defined the default monitor for vm is "qmp", so the code here is not needed.
> In additional, we are not sure if bitmap-add related cmds will be supported by hmp in the future, so i don't think it's suitable to add the limitation here.

1. VM's monitor type is reset by our internal configuration, so it is better to limit this case only for VMs with qmp support.
2. Also, I believe hmp will be deprecated in the future.

**Comment by lolyu** [comment]
```
 (1/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (142.03 s)
 (2/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (144.77 s)
 (3/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (138.37 s)
 (4/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (140.52 s)
 (5/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (140.50 s)

```

File: qemu/tests/live_backup_add_bitmap.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                                                         existence):
+            msg = "bitmap %s %s exists after addition" % (
+                bitmap, "not" if existence else "")
+            msgs.append(msg)
+    if msgs:
+        test.fail("\n".join(msgs))
+
+    # check bitmap existence after shutdown
+    error_context.context("system powerdown", logging.info)
+    vm.monitor.system_powerdown()
```
Comment (hereischen) [inline qemu/tests/live_backup_add_bitmap.py]: After line 59, the if nothing goes wrong the guest should be off.
So for line 60, you can just use `vm.is_dead()`?

Comment (lolyu) [reply qemu/tests/live_backup_add_bitmap.py]: `vm.wait_for_shutdown` just wait until guest shuts down, it doesn't send the actual shutdown command I believe.

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

**Comment by lolyu** [comment]
```
 (1/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (142.87 s)
 (2/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (130.98 s)
 (3/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (142.57 s)
 (4/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (141.84 s)
 (5/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (139.56 s)
```

**Comment by aliang123** [comment]
From  qemu/tests/live_backup_add_bitmap.py:
    line 49: for bitmap in params.get("bitmaps", "").split():
    we can see, the code support multiple bitmap add.
But in qemu/tests/cfg/live_backup_add_bitmap.cfg:
    The bitmap check is limited to only one bitmap.
So i think we should do some updates for  the multiple bitmap_add scenario.

**Comment by lolyu** [comment]
> From qemu/tests/live_backup_add_bitmap.py:
> line 49: for bitmap in params.get("bitmaps", "").split():
> we can see, the code support multiple bitmap add.
> But in qemu/tests/cfg/live_backup_add_bitmap.cfg:
> The bitmap check is limited to only one bitmap.
> So i think we should do some updates for the multiple bitmap_add scenario.

I believe  `check_bitmap_existence_as_expected` supports the scenario you mentioned.

**Comment by aliang123** [comment]
@lolyu, yes, your script support multiple bitmaps-add but your cfg not. If need to check your multiple bitmaps-add part, the cfg should be convenient to configure.

**Comment by lolyu** [comment]
```
 (1/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (87.96 s)
 (2/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (88.64 s)
 (3/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (89.45 s)
 (4/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (88.67 s)
 (5/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (89.55 s)
```

**Comment by lolyu** [comment]
```
(1/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (78.08 s)
 (2/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (73.26 s)
 (3/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (76.46 s)
 (4/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (76.01 s)
 (5/5) Host_RHEL.m7.u7.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (76.49 s)
```

**Comment by lolyu** [comment]
@aliang123 @hereischen updated, please take a review.

**Comment by aliang123** [comment]
RHEL7 + two bitmaps add:
    (1/5) Host_RHEL.m7.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (214.15 s)
 (2/5) Host_RHEL.m7.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (210.46 s)
 (3/5) Host_RHEL.m7.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (210.32 s)
 (4/5) Host_RHEL.m7.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (205.52 s)
 (5/5) Host_RHEL.m7.u7.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (212.24 s)

RHEL8+one bitmap add:
   (1/5) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_on: PASS (74.54 s)
 (2/5) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_off: PASS (72.19 s)
 (3/5) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_qcow2.persistent_default: PASS (72.24 s)
 (4/5) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_off: PASS (74.00 s)
 (5/5) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.1.x86_64.io-github-autotest-qemu.live_backup_add_bitmap.with_raw.persistent_default: PASS (72.85 s)

Acked-by: Aihua Liang <aliang@redhat.com>

**Comment by lolyu** [comment]
Thanks y'all.

**Comment by aliang123** [comment]
id:1824051

**Pull request merged**